core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java

```diff
@@ -549,19 +549,44 @@ private long processWalSql(TableWriter tableWriter, WalEventCursor.SqlInfo sqlIn
         operationExecutor.resetRnd(sqlInfo.getRndSeed0(), sqlInfo.getRndSeed1());
         sqlInfo.populateBindVariableService(operationExecutor.getBindVariableService());
         try {
-            switch (cmdType) {
-                case CMD_ALTER_TABLE:
-                    operationExecutor.executeAlter(tableWriter, sql, seqTxn);
-                    return -1;
-                case CMD_UPDATE_TABLE:
-                    return operationExecutor.executeUpdate(tableWriter, sql, seqTxn);
-                default:
-                    throw new UnsupportedOperationException("Unsupported command type: " + cmdType);
+            while (true) {
+                try {
+                    switch (cmdType) {
+                        case CMD_ALTER_TABLE:
+                            operationExecutor.executeAlter(tableWriter, sql, seqTxn);
+                            return -1;
+                        case CMD_UPDATE_TABLE:
+                            return operationExecutor.executeUpdate(tableWriter, sql, seqTxn);
+                        default:
+                            throw new UnsupportedOperationException("Unsupported command type: " + cmdType);
+                    }
+                } catch (SqlException ex) {
+                    if (!ex.isTableDoesNotExist()) {
+                        throw ex;
+                    }
+                } catch (CairoException ex) {
+                    if (!ex.isTableDoesNotExist()) {
+                        throw ex;
+                    }
+                }
+
+                TableToken tableToken = tableWriter.getTableToken();
+
+                // Here means we got TableDoesNotExist SQL or Cairo Exception.
+                // Table may be renamed while processing the WAL transaction.
+                // Need to refresh the table token and retry.
+                TableToken updatedToken = engine.getUpdatedTableToken(tableToken);
```

Comment:
can we refresh table token unconditionally before executing update, or is it too expensive to do upfront?

Reply:
We can refresh beforehand, but this code will have to stay.
Refreshing is only needed after renames.